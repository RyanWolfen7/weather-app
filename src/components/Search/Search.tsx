import React, { useCallback, useState } from 'react';
import { withRouter, RouteComponentProps } from 'react-router';
import { InputGroup, InputGroupAddon, Button, Input, Form } from 'reactstrap';
import styled from 'styled-components';
import { useDispatch } from 'react-redux';
import actions from '../../actions';

interface SearchInterface extends RouteComponentProps {
  dir?: string;
}

const Search: React.FC<SearchInterface> = props => {
  const [city, setCity] = useState('');
  const dispatch = useDispatch();

  const onCityChange = useCallback(e => {
    setCity(e.target.value);
  }, []);

  const preventSubmit = useCallback(e => e.preventDefault(), []);

  const getWeather = () => {
    dispatch(actions.weather.clearWeather());
    props.history.push(`/forecast/${city}`);
  };

  return (
    <Wrapper>
      <Form onSubmit={preventSubmit}>
        <InputWrapper dir={props.dir}>
          <Input placeholder={'St. George, Utah'} value={city} valid={city.length > 2} onChange={onCityChange} />
          <InputGroupAddon addonType="append">
            <Button onClick={getWeather} color="success" disabled={city.length < 3} type={'submit'}>Get Weather</Button>
          </InputGroupAddon>
        </InputWrapper>
      </Form>
    </Wrapper>
  );
}

export default withRouter(Search);

const Wrapper = styled.div`
  width: 100%;
  max-width: 300px;
`;

const InputWrapper = styled(InputGroup)`
  flex-direction: ${props => props.dir || 'row'};
  justify-content: space-around!important;
  align-items: center!important;
`;
