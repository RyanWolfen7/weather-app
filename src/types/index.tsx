export const FETCH_WEATHER = '@WEATHER/FETCH_WEATHER';
export const FETCH_WEATHER_SUCCESS = '@WEATHER/FETCH_WEATHER_SUCCESS';
export const FETCH_WEATHER_FAILURE = '@WEATHER/FETCH_WEATHER_FAILURE';
export const CLEAR_WEATHER = '@WEATHER/CLEAR_WEATHER';

enum Types {
  FETCH_WEATHER,
  FETCH_WEATHER_SUCCESS,
  FETCH_WEATHER_FAILURE,
  CLEAR_WEATHER,
};

export default Types;
