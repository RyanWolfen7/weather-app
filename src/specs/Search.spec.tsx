import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import Search from '../components/Search/Search';
import { CLEAR_WEATHER } from '../types';

let wrapper: ReactWrapper;
let location = { pathname: '/' };
const useDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => useDispatch,
}));

beforeAll(() => {
  wrapper = mount(
    <MemoryRouter initialEntries={['/']}>
      <Search />
      <Route
        path="*"
        render={props => {
          location = props.location;
          return null;
        }}
      />
    </MemoryRouter>
  );
});

test('Search the weather by city name', () => {
  wrapper.find('input').simulate('change', { target: { value: 'testing' } });
  wrapper.find('button').simulate('click');
  expect(useDispatch).toBeCalledWith({ type: CLEAR_WEATHER });
  expect(location.pathname).toEqual('/forecast/testing');
});

test('Search to match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});
