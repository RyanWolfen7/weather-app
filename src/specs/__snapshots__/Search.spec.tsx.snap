// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search to match snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "PUSH",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          Object {
            "hash": "",
            "pathname": "/forecast/testing",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 1,
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/forecast/testing",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(Search)>
      <Search
        history={
          Object {
            "action": "PUSH",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              Object {
                "hash": "",
                "pathname": "/forecast/testing",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 1,
            "length": 2,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/forecast/testing",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/forecast/testing",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      >
        <styled.div>
          <div
            className="sc-AykKC iwRJSO"
          >
            <Form
              onSubmit={[Function]}
              tag="form"
            >
              <form
                className=""
                onSubmit={[Function]}
              >
                <Styled(InputGroup)>
                  <InputGroup
                    className="sc-AykKD kDqypH"
                    tag="div"
                  >
                    <div
                      className="sc-AykKD kDqypH input-group"
                    >
                      <Input
                        onChange={[Function]}
                        placeholder="St. George, Utah"
                        type="text"
                        valid={true}
                        value="testing"
                      >
                        <input
                          className="is-valid form-control"
                          onChange={[Function]}
                          placeholder="St. George, Utah"
                          type="text"
                          value="testing"
                        />
                      </Input>
                      <InputGroupAddon
                        addonType="append"
                        tag="div"
                      >
                        <div
                          className="input-group-append"
                        >
                          <Button
                            color="success"
                            disabled={false}
                            onClick={[Function]}
                            tag="button"
                            type="submit"
                          >
                            <button
                              aria-label={null}
                              className="btn btn-success"
                              disabled={false}
                              onClick={[Function]}
                              type="submit"
                            >
                              Get Weather
                            </button>
                          </Button>
                        </div>
                      </InputGroupAddon>
                    </div>
                  </InputGroup>
                </Styled(InputGroup)>
              </form>
            </Form>
          </div>
        </styled.div>
      </Search>
    </withRouter(Search)>
    <Route
      path="*"
      render={[Function]}
    />
  </Router>
</MemoryRouter>
`;
